File: logbook-spring-webflux/src/main/java/org/zalando/logbook/spring/webflux/ClientRequest.java
package org.zalando.logbook.spring.webflux;

import lombok.RequiredArgsConstructor;
import org.springframework.util.MimeType;
import org.zalando.logbook.HttpHeaders;
import org.zalando.logbook.HttpRequest;
import org.zalando.logbook.Origin;
import org.zalando.logbook.attributes.HttpAttributes;

import javax.annotation.Nullable;
import java.io.IOException;
import java.net.URI;
import java.nio.charset.Charset;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

import static java.nio.charset.StandardCharsets.UTF_8;

@RequiredArgsConstructor
final class ClientRequest implements HttpRequest {

    private final org.springframework.web.reactive.function.client.ClientRequest request;

    private final AtomicReference<State> state = new AtomicReference<>(new State.Unbuffered());

    @Override
    public String getRemote() {
        return "localhost";
    }

    @Override
    public String getMethod() {
        return request.method().name();
    }

    @Override
    public String getScheme() {
        return Optional.of(request.url())
                .map(URI::getScheme)
                .orElse("");
    }

    @Override
    public String getHost() {
        return Optional.of(request.url())
                .map(URI::getHost)
                .orElse("");
    }

    @Override
    public Optional<Integer> getPort() {
        return Optional.of(request.url().getPort())
                .filter(p -> p != -1);
    }

    @Override
    public String getPath() {
        return Optional.of(request.url())
                .map(URI::getPath)
                .orElse("");
    }

    @Override
    public String getQuery() {
        return Optional.of(request.url())
                .map(URI::getQuery)
                .orElse("");
    }

    @Override
    public Origin getOrigin() {
        return Origin.LOCAL;
    }

    @Override
    public HttpHeaders getHeaders() {
        return HttpHeaders.of(request.headers());
    }

    @Nullable
    @Override
    public String getContentType() {
        return Optional.ofNullable(request.headers().getContentType())
                .map(MimeType::toString)
                .orElse(null);
    }

    @Override
    public Charset getCharset() {
        return Optional.ofNullable(request.headers().getContentType())
                .map(MimeType::getCharset)
                .orElse(UTF_8);
    }

    @Override
    public HttpRequest withBody() throws IOException {
        state.updateAndGet(State::with);
        return this;
    }

    @Override
    public HttpRequest withoutBody() {
        state.updateAndGet(State::without);
        return this;
    }

    boolean shouldBuffer() {
        return state.get() instanceof State.Offering;
    }

    void buffer(byte[] message) {
        state.updateAndGet(s -> s.buffer(message));
    }

    @Override
    public byte[] getBody() throws IOException {
        return state.get().getBody();
    }

    @Override
    public HttpAttributes getAttributes() {
        return new HttpAttributes(request.attributes());
    }
}

File: logbook-ktor-client/src/main/kotlin/org/zalando/logbook/client/ClientRequest.kt
@file:Suppress(
    "SimpleRedundantLet", // jacoco workaround
)

package org.zalando.logbook.client

import io.ktor.client.request.HttpRequestBuilder
import io.ktor.client.request.host
import io.ktor.client.request.port
import io.ktor.http.HttpProtocolVersion.Companion.HTTP_1_1
import io.ktor.http.charset
import io.ktor.http.content.OutgoingContent
import io.ktor.http.contentType
import io.ktor.http.encodedPath
import io.ktor.util.toMap
import kotlin.text.Charsets.UTF_8
import org.zalando.logbook.ContentType.CONTENT_TYPE_HEADER
import org.zalando.logbook.HttpHeaders
import org.zalando.logbook.HttpRequest
import org.zalando.logbook.Origin
import org.zalando.logbook.common.State
import java.nio.charset.Charset
import java.util.Optional
import java.util.concurrent.atomic.AtomicReference

internal class ClientRequest(
    private val request: HttpRequestBuilder,
) : HttpRequest {
    private val state: AtomicReference<State> = AtomicReference(State.Unbuffered)

    override fun getProtocolVersion(): String = HTTP_1_1.toString()
    override fun getOrigin(): Origin = Origin.LOCAL
    override fun getHeaders(): HttpHeaders = HttpHeaders.of(request.headers.build().toMap()).also {
        // as Ktor removes the Content-Type header from the headers in the default transformers we need to add it back
        // https://github.com/ktorio/ktor/blob/e789b015bd5169505df1c59ced0d2690026af523/ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/DefaultTransform.kt#L55
        if (it.contains(CONTENT_TYPE_HEADER).not() && request.body is OutgoingContent) return it.update(
            CONTENT_TYPE_HEADER, (request.body as OutgoingContent).contentType?.toString()
        )
    }

    override fun getContentType(): String? = (request.contentType() ?: (request.body as? OutgoingContent)?.contentType)
            ?.let { it.toString().substringBefore(";") }
    override fun getCharset(): Charset = request.charset() ?: UTF_8
    override fun getRemote(): String = "localhost"
    override fun getMethod(): String = request.method.value
    override fun getScheme(): String = request.url.protocol.name
    override fun getHost(): String = request.host
    override fun getPort(): Optional<Int> = Optional.of(request.port)
    override fun getPath(): String = request.url.encodedPath
    override fun getQuery(): String = request.url.buildString().substringAfter("?", "")
    override fun withBody(): HttpRequest = apply { state.updateAndGet { it.with() } }
    override fun withoutBody(): HttpRequest = apply { state.updateAndGet { it.without() } }
    override fun getBody(): ByteArray = state.get().body
    internal fun buffer(bytes: ByteArray): State = state.updateAndGet { it.buffer(bytes) }
    internal fun shouldBuffer(): Boolean = state.get() is State.Offering
}
